### DISCUSSION TITLE
Questions about evaluating expressions and field symbol discarding/generation.

### DISCUSSION THREAD

**Discussion by AddioElectronics** [dotnet/roslyn#71628] [Q&A] [answered]
I've been adding a new feature, and ran into a little problem.
The new syntax I have added has a constant expression that can't be bound to anything, but I need to calculate the value of the expression during compile time.
Is there any way to evaluate an expression without binding anything?

On another note, if its possible, how can I discard members that I don't need, specifically backing fields?
I saw the comment below which lead me to believe its possible, but I have not found any specific code that does yet.
Or instead of discarding unused ones, is it possible to add members (fields to a struct) later in compilation, like during method synthesizing? 

> This should not attempt to bind any method parameters as that would cause
the members being built to be captured in the binder cache before the final
list of members is determined.

**Comment by CyrusNajmabadi** [comment]
Could you provide an example of what you are trying to do?

**Comment by AddioElectronics** [comment]
Just something like this, called from inside my symbol after ForceComplete is finished.
Its basically an initializer that isn't bound to a field.

```
ExpressionSyntax expression = syntax.Declaration.Expression;
SemanticModel semanticModel = compilation.GetSemanticModel(syntax.SyntaxTree);
var value = semanticModel.GetConstantValue(expression);
```

GetConstantValue() ends up in SyntaxTreeSemanticModel.GetMemberModel(), where I have started to add my syntax.
But from what I can tell looking at the EqualsValueClause, these steps are for binding the expression to a field, which I'd like to be able to skip.

And then once I get the value, I may or may not require a backing field, which is why I either want to discard the backing field that was created with the symbol, or create a new one during method synthesizing(And then insert in to the struct member list).

**Reply by CyrusNajmabadi** [reply]
not really getting it.  can you give an example of hwat the code would be that you would be compiling, and what semantics you want it to have?

**Reply by AddioElectronics** [reply, marked as answer]
Probably because I didn't really get it either. Turns out GetConstantValue did something different than what I thought.

I ended up figuring it out yesterday.

```
var binderFactory = compilation.GetBinderFactory(syntax.SyntaxTree);
var binder = binderFactory.GetBinder(expression);

BoundExpression boundExpression = binder.BindValue(expression, diagnostics, Binder.BindValueKind.RValue);

return boundExpression.ConstantValueOpt;
```

I think I found out the backing field problem as well.
Inside MethodCompiler.CompileNamedType() the PassesFilter() function is called, will have to look into it tomorrow though.

> can you give an example of what the code would be that you would be compiling, and what semantics you want it to have?

I've been adding bitfields, and was trying to get the length.
Just got it all working, compiles and runs, now I just need to figure out how to get rid of some of the backing fields.

Edit: Looks like PassesFilter isn't an option.

Edit2: Figured it out, I can skip adding the backing field to the member list, and generate as needed.
Then inside SourceMemberContainerSymbol.GetFieldsToEmit(), I just need to return the backing field from my symbol.
